crimson/common/errorator: rework aborts error handlers

Matan-B opened the pull request:
This came up during: https://tracker.ceph.com/issues/69406#note-25
Where an "assert_all" was called but didn't cause an abort.

This PR is a rework of errorator aborting paths with fixes and some cleanups.


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

Matan-B commented on src/test/crimson/test_errorator_abort.cc:
```diff
@@ -0,0 +1,105 @@
+{
+  run_async([this] {
+    return invarg_foo().handle_error(
+      crimson::ct_error::assert_all("unexpected error")
```
The issue is that we expect assert_all to call abort internally. When it's not ignored we are throwing an exception. However, when ignored it will only result in a warning (instead of an abort).
@rzarzynski, I'm looking into this and was wondering if you have any suggestions/insights on this one.

Matan-B replied:
Warning is:
```
WARN  2025-04-07 09:01:29,690 [shard  0:main] seastar - Exceptional future ignored: crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<22> >::throwable_carrier, backtrace: /lib
64/libasan.so.8+0x6bddd 0x17cc05a 0x17cc830 0x17cd909 0x17ce2ea 0xfd02e5 0xfd065a 0x6f4271 0x6ff730 0x72d2d3 0x75f3a1 0x1150509 0x115b79b 0x1213d39 0x12151c7 0xf32b07 0xf33317 0xf334c9 0x789
1c5 0x7895a0 0x78973f /lib64/libstdc++.so.6+0xdbae3 /lib64/libc.so.6+0x89c01 /lib64/libc.so.6+0x10ec3f
   --------                                                                                    
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>, seastar::future<void>::then_wrapped_
nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&&)
::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&, seastar::future_state<seastar::internal::monosta
te>&&)#1}, void>
```

Matan-B replied:
Thanks Radek, I've reverted https://github.com/ceph/ceph/commit/513257cdd67ec126539f617256d73e4e8c65cf31 for now. This seems to work - meaning, we abort successfully in any case shown in the test added.
I'm trying to understand why we've disallowed void returns https://github.com/ceph/ceph/pull/55735 in the first place. CC @xxhdx1985126.

cyx1231st replied:
I recall that `no_touch_error_marker` was supposed to be introduced **only** as a special label for `assert_all` and `assert_failure`, so that we can enforce the other pathes to always have non-void valid return types.

Is it because `errfunc` is mistakenly skipped in case of `no_touch_error_marker`, so that the `ceph_abort(...)` are mistakenly skipped in `assert_all::operator()` and `assert_failure::operator()` ?

xxhdx1985126 replied:
I think the issue is that the `composer` shouldn't return `no_touch_error_marker`. Will go deeper.

Matan-B replied:
Please see the test cases added. The abort is treated as a thrown exception and  if ignored we we'll never abort.

xxhdx1985126 replied:
> @xxhdx1985126 Please see `(errorator_abort_test_t, assert_all / ignore_assert_all / ignore_assert_failure)` **non** of these are aborting.

I know, but I don't think it is caused by `no_touch_error_marker`, because all the methods that are returning `no_touch_error_marker` are actually calling `abort()` before returning. `no_touch_error_marker` is just a label.

> 
> After reverting all of them are aborting successfully as expected. Should we move forward with the revert or is the `no_touch_error_marker` concept still needed?

That commit is needed to fix https://tracker.ceph.com/issues/64556

xxhdx1985126 replied:
Ah, my mistake, I think we can call the handler before returning if it is returning `no_touch_error_marker`.

Matan-B replied:
> That commit is needed to fix https://tracker.ceph.com/issues/64556.

I'm not sure I understand the issue there, shouldn't we change the user (in this case `OMapLeafNode::get_value: key = omap_header`) to handle the issue correctly?

>Ah, my mistake, I think we can call the handler before returning if it is returning no_touch_error_marker.

Can you please show what do you mean?

xxhdx1985126 replied:
> > That commit is needed to fix https://tracker.ceph.com/issues/64556.
> 
> I'm not sure I understand the issue there, shouldn't we change the user (in this case `OMapLeafNode::get_value: key = omap_header`) to handle the issue correctly?

No, that issue was caused by not properly processs a previous error.
> 
> > Ah, my mistake, I think we can call the handler before returning if it is returning no_touch_error_marker.
> 
> Can you please show what do you mean?

Something like this:
```
diff --git a/src/crimson/common/errorator.h b/src/crimson/common/errorator.h
index 5d30d6f14d4..e05a493ea51 100644
--- a/src/crimson/common/errorator.h
+++ b/src/crimson/common/errorator.h
@@ -369,7 +369,10 @@ public:
     using return_t = std::invoke_result_t<ErrorVisitorT, ErrorT>;
     static_assert(!std::is_same_v<return_t, void>,
                   "error handlers mustn't return void");
+    [[maybe_unused]] auto &&ep = std::move(result).get_exception();
     if constexpr (std::is_same_v<return_t, no_touch_error_marker>) {
+      result = std::invoke(std::forward<ErrorVisitorT>(errfunc),
+                           ErrorT::error_t::from_exception_ptr(std::move(ep)));
       return;
     } else {
       // In C++ throwing an exception isn't the sole way to signal
@@ -394,7 +397,6 @@ public:
       if (type_info == ErrorT::error_t::get_exception_ptr_type_info()) {
         // set `state::invalid` in internals of `seastar::future` to not
         // call `report_failed_future()` during `operator=()`.
-        [[maybe_unused]] auto &&ep = std::move(result).get_exception();
         if constexpr (std::is_assignable_v<decltype(result), return_t>) {
           result = std::invoke(std::forward<ErrorVisitorT>(errfunc),
                                ErrorT::error_t::from_exception_ptr(std::move(ep)));
```

Matan-B replied:
> @Matan-B Something like this should fix the issue:
> 
> ```
>      if constexpr (std::is_same_v<return_t, no_touch_error_marker>) {
> +      result = std::invoke(std::forward<ErrorVisitorT>(errfunc),
> +                           ErrorT::error_t::from_exception_ptr(std::move(ep)));
>        return;
> ```

That won't work, any handle_error user will FTBFS:
```
/ceph/src/test/crimson/test_errorator_abort.cc:73:37:   required from here                                                                                                         
/ceph/src/crimson/common/errorator.h:374:14: error: no match for ‘operator=’ (operand types are ‘seastar::futurize<seastar::future<> >::type’ {aka ‘seastar::future<>’} and ‘std::i
nvoke_result_t<crimson::ct_error::assert_all, const crimson::unthrowable_wrapper<const std::error_code&, ((const std::error_code&)(& crimson::ec<22>))>&>’ {aka ‘crimson::no_touch_error_marke
r’})                                                                                                                                                                                          
  374 |       result = std::invoke(std::forward<ErrorVisitorT>(errfunc),
      |       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 375  |                            ErrorT::error_t::from_exception_ptr(std::move(ep)));  
```
Please see my previous comment

xxhdx1985126 replied:
@Matan-B This should fix the issue: https://github.com/xxhdx1985126/ceph/commit/52e2ce5d0b2060bbcfbed1a8f5ad317406d934e1

rzarzynski replied:
What's the rationale for bypassing the `type_info == ErrorT::error_t::get_exception_ptr_type_info()` check?

cyx1231st replied:
If I remember correctly, see an example https://github.com/ceph/ceph/pull/55735/commits/2f2b4f6170dbbd24c404d2bf58170e0818df8c80#diff-9bf352bd708e64bbcf2c4a3adf2dbf3c1443acde26cebab25bf764a890dd62e6L303-L307

If there is no `no_touch_error_marker`, we were allowed to write error handlers mistakenly returning void (where it is supposed to return a future), that caused segmentation fault such as https://tracker.ceph.com/issues/64556. It was error-prone, and `no_touch_error_marker` is helpful to prevent them at compile time.

xxhdx1985126 replied:
> Thank you @xxhdx1985126, I've updated the PR with the commit above - it seems to work well.
> 
> If we decide to go with this fix instead of reverting, I think that we should have proper explanation about the `no_touch_error_marker` concept and how it solves issues like https://tracker.ceph.com/issues/64556 by disallowing void returns (since this is still not fully clear to me). What do you think?

As discussed in https://github.com/ceph/ceph/pull/55735#discussion_r1502130462, the problem was that the void-returning error handler would pass on the exception future instead of consuming it, although the errfunc is called, which led to the issue https://tracker.ceph.com/issues/64556 through a chain of events.

So the idea was to ban the void-returning handler. However, the `assert_xxx` handler in errorator were all returning void, so we added `no_touch_error_marker` to label those methods, forcing them to also return something instead of void.

> What's the rationale for bypassing the `type_info == ErrorT::error_t::get_exception_ptr_type_info()` and the entire idea behind `no_touch_error_marker`?

I think this was a mistake, will correct it.

xxhdx1985126 replied:
> @Matan-B This should fix the issue: [xxhdx1985126@52e2ce5](https://github.com/xxhdx1985126/ceph/commit/52e2ce5d0b2060bbcfbed1a8f5ad317406d934e1)

@Matan-B I've just fixed the issue that @rzarzynski brought about and force pushed this commit, please take a look:-)

liu-chunmei reviewed (approved):
LGTM.

liu-chunmei commented:
jenkins test make check

liu-chunmei commented:
jenkins test make check arm64

liu-chunmei commented:
jenkins test make check

Matan-B commented:
jenkins test make check arm64

Matan-B commented:
changeset (last commits):
Pulled https://github.com/ceph/ceph/pull/62868
Removed `pre_assert` concept which was used solely yo obtain the error info
Added verbosity to any existing assert_all in the osd files
Added error printing to assert_failure as well

athanatos commented:
@Matan-B Above run looks good.

cyx1231st commented on src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_accessor.h:
```diff
@@ -551,21 +548,12 @@ class NodeExtentAccessorT {
+          "{} EIO during retire -- to_disgard={}, fresh={}",
+          FNAME, to_discard->get_laddr(), fresh_extent->get_laddr()).c_str()),
+        crimson::ct_error::enoent::assert_failure(fmt::format(
+          "{} EIO during retire -- to_disgard={}, fresh={}",
```
nit, EIO -> ENOENT.

If the error handler can print the error type, the more general `assert_all()` can be used.

Matan-B replied:
> If the error handler

It can now :)
Missed the assert_all usecase - updated!

cyx1231st commented on src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_accessor.h:
```diff
@@ -580,14 +568,11 @@ class NodeExtentAccessorT {
+      crimson::ct_error::input_output_error::assert_failure(fmt::format(
+        "{} EIO -- addr={}", FNAME, addr).c_str()),
+      crimson::ct_error::enoent::assert_failure(fmt::format(
+        "{} EIO -- addr={}", FNAME, addr).c_str()
```
nit, EIO -> ENOENT.

If the error handler can print the error type, the more general `assert_all()` can be used.

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test api

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test api

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test api

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test make check arm64

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test api

Matan-B commented:
jenkins test api

Matan-B commented:
jenkins test make check arm64

Matan-B commented:
Merging based on https://pulpito.ceph.com/matan-2025-04-23_07:18:01-crimson-rados-wip-matanb-crimson-ignore-abort-v2-distro-crimson-smithi/

Matan-B commented:
jenkins test make check arm64

Matan-B commented:
Unreleated:
`	 30 - run-rbd-unit-tests-0.sh (Failed`
